quint/src/effects/modeChecker.ts

```diff
@@ -107,12 +129,19 @@ class ModeFinderVisitor implements EffectVisitor {
 
     const [paramReads, paramUpdates, paramTemporals] = paramVariablesByEffect(effect)
 
-    if (anyVariablesAdded(paramTemporals, r.temporal)) {
+    const addedTemporal = addedVariables(paramTemporals, r.temporal)
+    const addedUpdates = addedVariables(paramUpdates, r.update)
+    const addedReads = addedVariables(paramReads, r.read)
+
+    if (addedTemporal.length > 0) {
       this.currentMode = 'temporal'
-    } else if (anyVariablesAdded(paramUpdates, r.update)) {
+      this.explanation = `performs temporal operations over variables ${addedTemporal.map(variablesToString)}`
+    } else if (addedUpdates.length > 0) {
       this.currentMode = 'action'
-    } else if (anyVariablesAdded(paramReads, r.read)) {
+      this.explanation = `updates variables ${addedUpdates.map(variablesToString)}`
+    } else if (addedReads.length > 0) {
       this.currentMode = 'def'
+      this.explanation = `reads variables ${addedReads.map(variablesToString)}`
```

Comment:
Why do with these explanations need to be set by a mutable variable? In both of these two blocks where the explanations are formed, the text of the explanation is repeated verbatim and, iiuc, each message appears to be a function of two things:

- the current mode (e.g., `temporal`)
- the effect (from which the relevant set of variables are derived)

It looks to me as if both of these values are available at the point at which the final error message is constructed on line 68, so I'm wondering why we cannot computer along with the `modeConstraint`. On the other hand if we do need an additional mutable very able to carry some information here, it looks like we only really need the  set of variables.

WDYT? Am I missing something, or could this be cleaned up a bit?

Reply:
I was writing a big response to this, but I actually think that we don't even need the visitor with the mutable `currentMode` at all. I've changed the behavior of this since I first wrote it and now it seems like we don't need to look at nested effects anymore, only the top-level definition effect :smile: that will simplify things a lot, thank you for the thoughtful observation.

Reply:
oh very nice! Look forward to seeing the new code :)

Reply:
It's much better now! I didn't extract the explanation to a different function because I feel like when I read the explanation along with the logic, one thing helps understanding the other.

Reply:
makes sense1